File: scripts/build_tag_index.py
#!/usr/bin/env python3
"""
Build tag index from knowledge files.

Usage:
    python build_tag_index.py <knowledge_path>
    python build_tag_index.py <knowledge_path> --output <output_path>

Output:
    Creates tag-index.json in the knowledge directory (or specified output path).
"""

import json
import sys
import re
from pathlib import Path


def extract_frontmatter(content: str) -> dict:
    """Extract YAML frontmatter from markdown content."""
    if not content.startswith('---'):
        return {}

    lines = content.split('\n')
    end_index = -1
    for i, line in enumerate(lines[1:], 1):
        if line.strip() == '---':
            end_index = i
            break

    if end_index == -1:
        return {}

    frontmatter = {}
    current_key = None
    current_list = None

    for line in lines[1:end_index]:
        if not line.strip():
            continue

        # Check for key: value
        match = re.match(r'^(\w+):\s*(.*)$', line)
        if match:
            key = match.group(1)
            value = match.group(2).strip()

            # Handle inline list: tags: [a, b, c]
            if value.startswith('[') and value.endswith(']'):
                items = value[1:-1].split(',')
                frontmatter[key] = [item.strip() for item in items if item.strip()]
                current_key = None
                current_list = None
            elif value:
                frontmatter[key] = value
                current_key = None
                current_list = None
            else:
                # Start of a list
                current_key = key
                current_list = []
                frontmatter[key] = current_list
        elif current_list is not None and line.strip().startswith('-'):
            # List item
            item = line.strip()[1:].strip()
            current_list.append(item)

    # Filter out deprecated fields
    deprecated_fields = {'category', 'requires', 'provides'}
    return {k: v for k, v in frontmatter.items() if k not in deprecated_fields}


def build_tag_index(knowledge_path: Path) -> dict:
    """Build tag index from all markdown files in knowledge directory."""
    tags_map = {}  # tag -> list of file paths

    for md_file in knowledge_path.rglob('*.md'):
        # Skip tag-index.json itself and any non-knowledge files
        relative_path = md_file.relative_to(knowledge_path)
        path_str = str(relative_path).replace('\\', '/')

        # Read and parse frontmatter
        try:
            content = md_file.read_text(encoding='utf-8')
        except Exception:
            continue

        frontmatter = extract_frontmatter(content)
        tags = frontmatter.get('tags', [])
        file_id = frontmatter.get('id', md_file.stem)

        if not tags:
            continue

        # Build file path key (category/id.md format)
        category = str(relative_path.parent).replace('\\', '/') if relative_path.parent != Path('.') else 'root'
        file_key = f"{category}/{file_id}.md"

        for tag in tags:
            if tag not in tags_map:
                tags_map[tag] = []
            if file_key not in tags_map[tag]:
                tags_map[tag].append(file_key)

    # Sort tags and file lists for consistent output
    for tag in tags_map:
        tags_map[tag].sort()

    return {
        "tags": dict(sorted(tags_map.items()))
    }


def main():
    if len(sys.argv) < 2:
        print("Usage: python build_tag_index.py <knowledge_path> [--output <output_path>]", file=sys.stderr)
        sys.exit(1)

    knowledge_path = Path(sys.argv[1])

    if not knowledge_path.exists():
        print(f"Error: Knowledge path does not exist: {knowledge_path}", file=sys.stderr)
        sys.exit(1)

    # Parse optional output path
    output_path = knowledge_path / "tag-index.json"
    if "--output" in sys.argv:
        output_index = sys.argv.index("--output")
        if output_index + 1 < len(sys.argv):
            output_path = Path(sys.argv[output_index + 1])

    # Build index
    index = build_tag_index(knowledge_path)

    # Write output
    output_path.write_text(json.dumps(index, ensure_ascii=False, indent=2), encoding='utf-8')

    # Print success (no statistics)
    print(json.dumps({
        "status": "success",
        "output": str(output_path)
    }))


if __name__ == "__main__":
    main()

File: scripts/build-tag-index.js
#!/usr/bin/env node
/**
 * Build tag index from knowledge files.
 *
 * WHY: Creates searchable index for tag-based knowledge lookup.
 * Migrated from Python build_tag_index.py.
 *
 * Usage:
 *   node build-tag-index.js <knowledge_path>
 *   node build-tag-index.js <knowledge_path> --output <output_path>
 *
 * Output:
 *   Creates tag-index.json in the knowledge directory (or specified output path).
 */

const fs = require('fs');
const path = require('path');

// Deprecated fields to filter out from frontmatter
const DEPRECATED_FIELDS = new Set(['category', 'requires', 'provides']);

/**
 * Extract YAML frontmatter from markdown content.
 *
 * WHY: Frontmatter contains metadata (id, tags) needed for indexing.
 *
 * @param {string} content - Markdown file content
 * @returns {Object} Parsed frontmatter (excluding deprecated fields)
 */
function extractFrontmatter(content) {
    if (!content.startsWith('---')) {
        return {};
    }

    const lines = content.split('\n');
    let endIndex = -1;

    // Find closing ---
    for (let i = 1; i < lines.length; i++) {
        if (lines[i].trim() === '---') {
            endIndex = i;
            break;
        }
    }

    if (endIndex === -1) {
        return {};
    }

    const frontmatter = {};
    let currentKey = null;
    let currentList = null;

    for (let i = 1; i < endIndex; i++) {
        const line = lines[i];
        if (!line.trim()) {
            continue;
        }

        // Check for key: value
        const match = line.match(/^(\w+):\s*(.*)$/);
        if (match) {
            const key = match[1];
            const value = match[2].trim();

            // Handle inline list: tags: [a, b, c]
            if (value.startsWith('[') && value.endsWith(']')) {
                const items = value.slice(1, -1).split(',');
                frontmatter[key] = items
                    .map(item => item.trim())
                    .filter(item => item !== '');
                currentKey = null;
                currentList = null;
            } else if (value) {
                frontmatter[key] = value;
                currentKey = null;
                currentList = null;
            } else {
                // Start of a multiline list
                currentKey = key;
                currentList = [];
                frontmatter[key] = currentList;
            }
        } else if (currentList !== null && line.trim().startsWith('-')) {
            // List item
            const item = line.trim().slice(1).trim();
            currentList.push(item);
        }
    }

    // Filter out deprecated fields
    const result = {};
    for (const [key, value] of Object.entries(frontmatter)) {
        if (!DEPRECATED_FIELDS.has(key)) {
            result[key] = value;
        }
    }

    return result;
}

/**
 * Recursively find all markdown files in a directory.
 *
 * @param {string} dir - Directory to search
 * @param {string} [baseDir] - Base directory for relative paths
 * @returns {Array<{path: string, relativePath: string}>} List of file paths
 */
function findMarkdownFiles(dir, baseDir = dir) {
    const results = [];

    if (!fs.existsSync(dir)) {
        return results;
    }

    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);

        if (entry.isDirectory()) {
            results.push(...findMarkdownFiles(fullPath, baseDir));
        } else if (entry.isFile() && entry.name.endsWith('.md')) {
            const relativePath = path.relative(baseDir, fullPath).replace(/\\/g, '/');
            results.push({ path: fullPath, relativePath });
        }
    }

    return results;
}

/**
 * Build tag index from all markdown files in knowledge directory.
 *
 * WHY: Creates searchable index for tag-based knowledge lookup.
 *
 * @param {string} knowledgePath - Path to knowledge directory
 * @returns {Object} Tag index with { tags: { tagName: [files...] } }
 */
function buildTagIndex(knowledgePath) {
    const tagsMap = {}; // tag -> list of file paths

    const mdFiles = findMarkdownFiles(knowledgePath);

    for (const { path: filePath, relativePath } of mdFiles) {
        // Read and parse frontmatter
        let content;
        try {
            content = fs.readFileSync(filePath, 'utf-8');
        } catch (e) {
            continue;
        }

        const frontmatter = extractFrontmatter(content);
        const tags = frontmatter.tags || [];
        const fileId = frontmatter.id || path.basename(filePath, '.md');

        if (tags.length === 0) {
            continue;
        }

        // Build file path key (category/id.md format)
        const relativeDir = path.dirname(relativePath);
        const category = relativeDir === '.' ? 'root' : relativeDir.replace(/\\/g, '/');
        const fileKey = `${category}/${fileId}.md`;

        for (const tag of tags) {
            if (!tagsMap[tag]) {
                tagsMap[tag] = [];
            }
            if (!tagsMap[tag].includes(fileKey)) {
                tagsMap[tag].push(fileKey);
            }
        }
    }

    // Sort tags and file lists for consistent output
    for (const tag of Object.keys(tagsMap)) {
        tagsMap[tag].sort();
    }

    // Sort tags alphabetically
    const sortedTags = {};
    for (const key of Object.keys(tagsMap).sort()) {
        sortedTags[key] = tagsMap[key];
    }

    return {
        tags: sortedTags
    };
}

/**
 * Main CLI function.
 */
function main() {
    const args = process.argv.slice(2);

    if (args.length < 1) {
        console.error('Usage: node build-tag-index.js <knowledge_path> [--output <output_path>]');
        process.exit(1);
    }

    const knowledgePath = args[0];

    if (!fs.existsSync(knowledgePath)) {
        console.error(`Error: Knowledge path does not exist: ${knowledgePath}`);
        process.exit(1);
    }

    // Parse optional output path
    let outputPath = path.join(knowledgePath, 'tag-index.json');
    const outputIndex = args.indexOf('--output');
    if (outputIndex !== -1 && outputIndex + 1 < args.length) {
        outputPath = args[outputIndex + 1];
    }

    // Build index
    const index = buildTagIndex(knowledgePath);

    // Write output
    fs.writeFileSync(outputPath, JSON.stringify(index, null, 2), 'utf-8');

    // Print success (no statistics)
    console.log(JSON.stringify({
        status: 'success',
        output: outputPath
    }));
}

// Run main if executed directly
if (require.main === module) {
    main();
}

module.exports = {
    extractFrontmatter,
    buildTagIndex
};
